### ISSUE TITLE
Access schema include paths

### ISSUE THREAD

**Issue by justjacksonn** [raml-org/raml-js-parser#140]
Need to add the ability to get the actual !include path for schemas.

**Comment by dmartinezg** [comment]
Hi @justjacksonn I am not sure what this issue is about, can you clarify?

**Comment by blakeembrey** [comment]
Hi @dmartinezg! I asked him to log this and it was about being able to get the original `!include` path from a schema lookup. For example, `schema: !include schemas/user.json` loses the reference to `user.json` file and we can no longer do things with the original file (including validating that it's `$ref`s work, etc).

**Comment by mbixby** [comment]
+1

**Comment by blakeembrey** [comment]
Side note: I think the best way to keep backwards compatibility with the current parser would be to have the parser expand JSON schemas inline. It wouldn't require much of a change to support either.

**Comment by mbixby** [comment]
Thanks for the links, I think I'll just override `fetchLocalFileAsync` in `raml.coffee` to get it working until there's a final decision.

By the way, with this kind of preprocessing the underlying issue is still there, now you're not only hiding `!include` paths but also hiding `$ref` paths from the included schemas (mutating the schemas in the process).

I'd be okay if there was some new `schemaUri` attribute pointing to a virtual location that the user has to understand. You could then have some simple default method of resolving this virtual location into the full schema (like just using file paths) *and* provide option to override the behavior with something more advanced. Resolved schema would be still under the `schema` attribute.

**Comment by blakeembrey** [comment]
@mbixby I definitely understand and it's something that will be fixed in future parsers. Keeping backward compat right now doesn't give us much choice until the next major release which will break the current parser. I think this is the lesser of two evils, for now.

**Comment by maremmle** [comment]
+1

What's the current progress on this? I am currently struggling with schema `$ref` not getting resolved. Looks like https://github.com/bojand/json-schema-deref-sync (also mentioned in #155) is the perfect library to resolve this issue.

**Comment by przezor** [comment]
+1

**Comment by alvassin** [comment]
@blakeembrey +1 blocked by this issue . Have decided how to solve this issue?

**Comment by alvassin** [comment]
@martnst How you were able to achieve that outside of raml-js-parser?

**Comment by alvassin** [comment]
@dmartinezg @blakeembrey I have to keep my schemas in different folders, so this issue is very important for me. 

I can add some additional setting `resolveReferencedSchemas`, which would cause raml-js-parser to return text with substituted $ref to schemas. Is this solution acceptable?

I can make pull request with resolving referenced schemas without changing current interface of parser. Please let me know if it is that acceptable solution. If it is, i will make pull request.

**Comment by martnst** [comment]
I have a rather complex setup include forks of dependencies here and there. Basically I have a node script that wraps everything up. It's all in the context of [raml2hml](https://github.com/raml2html/raml2html). Therefor I am not sure my solution would help.

**Comment by blakeembrey** [comment]
@alvassin I do not work on this project anymore, @dmartinezg should be able to help you out. It sounds reasonable though.

**Comment by dmartinezg** [comment]
@alvassin, yes, if you can add the feature flag, we would be able to accept the patch.

**Comment by alvassin** [comment]
@dmartinezg I checked https://github.com/raml-org/raml-js-parser/pull/170 pull request. It is fine, except one issue: it does not deal with schemas, included at the root of RAML document. For inline inclusions it works fine.

Currently it's impossible to check if specified file belongs to `schemas` tag. We have just current & parent nodes, but we need parent.parent node. From my point of view, it would be nice to have resolved path for each node. Therefore we could easily say that current json file should be treated as json schema.

I found two calls [`descend_resolver`](https://github.com/raml-org/raml-js-parser/blob/master/src/composer.coffee#L99) and [`ascend_resolver`](https://github.com/raml-org/raml-js-parser/blob/master/src/composer.coffee#L106), that should do something similar, but they don't currently work because there are no `yaml_path_resolvers` defined.

I tried to search git history, but i can't find any examples of using `yaml_path_resolvers`. Perhaps you know something about this code (and can provide some advices / examples / docs / etc.)?

**Comment by dmartinezg** [comment]
I don't think it would be that simple to do while loading the document, because at that stage, the actual root `schema` key may not even be loaded, or may be included in another file. I think that this approach will fail for including root schemas.

Instead of doing this step in the YAML loading (in `raml.coffee`), it could be done as a composing step, in the `composeRamlTree` function at https://github.com/raml-org/raml-js-parser/blob/master/src/composer.coffee#L48 , where at that stage, you already have the YAML document read and validated, and can potentially access the entire AST by the root `node`.

The only remaining issue is handling the async nature of dereferencing the files, so maybe, the bit that `loads` the JSON files can remain at the async level, and the de-referencing can be done at the composing layer.

Makes sense @alvassin ?

**Comment by alvassin** [comment]
@dmartinezg Thanks for suggestion! I'll update code and add additional tests 👍

**Comment by alvassin** [comment]
@dmartinezg I implemented one more stage, that works after all fragment files are loade, but before composer (i would like to provide to composer ready object model for further validation & processing) - by adding [one more `then` callback](https://github.com/raml-org/raml-js-parser/blob/master/src/raml.coffee#L140).

Currently all cases are fine except inline defined schemas with `$ref`s: `process.cwd()` usually differs from raml file path, so  [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser) tries to find `$ref`erenced files in wrong place. 

E.g. for call `parser.load('tests/raml-files/expand-schemas.raml')`  `process.cwd()` will return `/Users/alvassin/Sites/raml-js-parser`, but raml file is located at `/Users/alvassin/Sites/raml-js-parser/tests/raml-files/expand-schemas.raml`

I found [issue](https://github.com/BigstickCarpet/json-schema-ref-parser/issues/13), where is said that json-schema-ref-parser itself does not provide ability to pass path for resolving `$ref`s. In this issue [it was suggested](https://github.com/BigstickCarpet/json-schema-ref-parser/issues/13#issuecomment-212439596) to use process.cwd as temporary solution.

I could change process.cwd (only) for the dereferencing stage, and then returning its value back. Is that acceptable?
